File: Strings/Palindromic_Substrings.java
class Palindromic_Substring {
    
    public int countPalindromicSubstring(String s) {

        int ans = 0;

        for(int i = 0; i < s.length(); i++) {
            // Odd length palindromes
            ans += expandAroundCenter(s, i, i);
            
            // Even length palindromes
            ans += expandAroundCenter(s, i, i+1);
        }

        return ans;

    }

    private int expandAroundCenter(String s, int L, int R) {
        
        int ans = 0;

        while(0 <= L && R < s.length() && s.charAt(L) == s.charAt(R)) {
            L--;
            R++;
            ans++;
        }

        return ans;
    }
}

// t: O(N^2), imagine this as there are N center candidates, each center can extend n-1 times. 
// s: O(1)

File: Python/2_String/Palindromic_Substrings.py
class Solution:
    def expandAroundCenter(self, s: str, l: int, r: int)->int:
        
        ans = 0
        while (0 <= l and r < len(s) and s[l] == s[r]):
            l -=1
            r += 1
            ans += 1
        
        return ans
    
    def countSubstrings(self, s: str) -> int:
        
        ans = 0
        for i in range(len(s)):
            # odd-amount palindromes
            ans += self.expandAroundCenter(s, i, i)
            # even-amount palindromes
            ans += self.expandAroundCenter(s, i, i+1)
            
        return ans
    
    # t: O(N ^ 2)
    # s: O(1)